multiversx_sdk/network_providers/transaction_awaiter.py

```diff
@@ -0,0 +1,90 @@
+    default_polling_interval = 6000
+    default_timeout = default_polling_interval * 15
+    default_patience = 0
+
+    def __init__(self,
+                 fetcher: ITransactionFetcher,
+                 polling_interval_in_milliseconds: Optional[int] = None,
+                 timeout_interval_in_milliseconds: Optional[int] = None,
+                 patience_time_in_milliseconds: Optional[int] = None) -> None:
+        self.fetcher = fetcher
+
+        if polling_interval_in_milliseconds is None:
+            self.polling_interval_in_milliseconds = TransactionAwaiter.default_polling_interval
+        else:
+            self.polling_interval_in_milliseconds = polling_interval_in_milliseconds
+
+        if timeout_interval_in_milliseconds is None:
+            self.timeout_interval_in_milliseconds = TransactionAwaiter.default_timeout
+        else:
+            self.timeout_interval_in_milliseconds = timeout_interval_in_milliseconds
+
+        if patience_time_in_milliseconds is None:
+            self.patience_time_in_milliseconds = TransactionAwaiter.default_patience
+        else:
+            self.patience_time_in_milliseconds = patience_time_in_milliseconds
+
+    def await_completed(self, tx_hash: str) -> TransactionOnNetwork:
+        """Waits until the transaction is completely processed."""
+        def is_completed(tx: TransactionOnNetwork):
+            if tx.is_completed is None:
+                raise IsCompletedFieldMissingOnTransaction()
+
+            return tx.is_completed
+
+        def do_fetch():
+            return self.fetcher.get_transaction(tx_hash)
+
+        return self._await_conditionally(
+            is_satisfied=is_completed,
+            do_fetch=do_fetch,
+            error=ExpectedTransactionStatusNotReached()
+        )
+
+    def _await_conditionally(self,
+                             is_satisfied: Callable[[TransactionOnNetwork], bool],
+                             do_fetch: Callable[[], TransactionOnNetwork],
+                             error: Exception) -> TransactionOnNetwork:
+        is_condition_satisfied = False
+        fetched_data: TransactionOnNetwork | None = None
+        max_number_of_retries = self.timeout_interval_in_milliseconds // self.polling_interval_in_milliseconds
+
+        number_of_retries = 0
+        while number_of_retries < max_number_of_retries:
+            try:
+                fetched_data = do_fetch()
+                is_condition_satisfied = is_satisfied(fetched_data)
+
+                if is_condition_satisfied:
+                    break
+            except Exception as ex:
+                raise ex
+
+            number_of_retries += 1
+            time.sleep(self.polling_interval_in_milliseconds / ONE_SECOND_IN_MILLISECONDS)
+
+        if fetched_data is None or not is_condition_satisfied:
+            raise error
+
+        if self.patience_time_in_milliseconds:
```

Comment:
what is the use case of patience time? even if the condition is satisfied there will still be a wait time if patience time is set

Reply:
maybe add some more info in the class docstring

Reply:
Added `docstring` and explained the usage of the `patience`.